src/predictor/cpu_predictor.cc

```diff
@@ -99,51 +99,77 @@ class AdapterView {
                         gbm::GBTreeModel const &model, int32_t tree_begin,
                         int32_t tree_end,
                         std::vector<RegTree::FVec> *p_thread_temp) {
   auto& thread_temp = *p_thread_temp;
+
   int32_t const num_group = model.learner_model_param->num_output_group;
 
   std::vector<bst_float> &preds = *out_preds;
   CHECK_EQ(model.param.size_leaf_vector, 0)
       << "size_leaf_vector is enforced to 0 so far";
   // parallel over local batch
   const auto nsize = static_cast<bst_omp_uint>(batch.Size());
-  auto constexpr kUnroll = DataView::kUnroll;
-  const bst_omp_uint rest = nsize % kUnroll;
-  if (nsize >= kUnroll) {
-#pragma omp parallel for schedule(static)
-    for (bst_omp_uint i = 0; i < nsize - rest; i += kUnroll) {
-      const int tid = omp_get_thread_num();
-      RegTree::FVec &feats = thread_temp[tid];
-      int64_t ridx[kUnroll];
-      SparsePage::Inst inst[kUnroll];
-      for (size_t k = 0; k < kUnroll; ++k) {
-        ridx[k] = static_cast<int64_t>(batch.base_rowid + i + k);
-      }
-      for (size_t k = 0; k < kUnroll; ++k) {
-        inst[k] = batch[i + k];
+
+  const size_t n_blocks = omp_get_max_threads();
+  const size_t block_size = nsize / n_blocks;
+  const size_t tail_size = nsize - n_blocks*block_size;
+  #pragma omp parallel num_threads(n_blocks)
+  {
+    const int tid = omp_get_thread_num();
+    const size_t begin = tid*block_size + batch.base_rowid;
+    size_t end = (tid == (n_blocks - 1)) ? nsize : (tid+1)*block_size;
+    end += batch.base_rowid;
+    const size_t n_cached_blocks = (end - begin) / block_of_rows_size;
+    const size_t tail_block_of_rows_size = (end  - begin) - block_of_rows_size*n_cached_blocks;
+
+    // process block of rows through all trees to keep cache locality
+    for (size_t block_id = 0; block_id < n_cached_blocks; ++block_id) {
+      const size_t offset = begin + block_id*block_of_rows_size;
+      for (size_t i = 0; i < block_of_rows_size; ++i) {
+        RegTree::FVec &feats = thread_temp[tid*block_of_rows_size + i];
+        const SparsePage::Inst inst = batch[offset + i - batch.base_rowid];
+        feats.Fill(inst);
       }
-      for (size_t k = 0; k < kUnroll; ++k) {
-        for (int gid = 0; gid < num_group; ++gid) {
-          const size_t offset = ridx[k] * num_group + gid;
-          preds[offset] += PredValue(inst[k], model.trees, model.tree_info, gid,
-                                     &feats, tree_begin, tree_end);
+      for (size_t tree_id = tree_begin; tree_id < tree_end; ++tree_id) {
+        const size_t gid = model.tree_info[tree_id];
+        for (size_t i = 0; i < block_of_rows_size; ++i) {
+          const int lid = thread_temp[tid*block_of_rows_size + i].HasMissing() ?  // 35% speed up
+          model.trees[tree_id]->GetLeafIndex<true>(thread_temp[tid*block_of_rows_size + i]) :
+          model.trees[tree_id]->GetLeafIndex<false>(thread_temp[tid*block_of_rows_size + i]);
```

Comment:
This contraction is really hard to read.
As I understand you want to avoid ```if``` here.
Can we compute both parts firstly and do ternar operation after? Or use __builtin_expect.

Reply:
it was moved to separate function, and was simplified significantly.